proposals/0165-async-vote-execution.md

```diff
@@ -0,0 +1,218 @@
+* vote account hash calculated above
+* block-id of the current bank
+
+This step is optimistic in the sense that validators do not check the fee
+payers when executing the vote transactions in a block. They assume vote
+transactions will not fail due to insufficient fees, apply the execution
+results to select the correct fork, then immediately vote on the bank with
+only the hash result of the vote transactions.
+
+This is safe because the list of validators and their corresponding stake
+uses the leader scheduler epoch stakes, which is calculated at the beginning
+of last Epoch. Because full execution is never behind the optimistic execution
+by more than one Epoch, the epoch stakes used is stable and correct.
+
+To make sure the vote casted would be the same as that after replaying the
+whole block, we need to be consistent on whether we mark the block dead, so
+that the ephemeral hash vote doesn't vote on a block which will be marked
+dead later. Currently a block can be dead for the following reasons:
+
+1. Unable to load data from blockstore
+2. Invalid block (wrong number of ticks, duplicate block, bad last fec, etc)
+3. Error while set root
+4. Invalid transaction
+
+For the first two, the same check can be performed computing ephemeral hash.
+We will set root on a bank only when it has full hash computed later, so the
+behavior will be the same as now.
+
+The only operation we can't check is invalid transaction, since we will skip
+all non-vote transaction execution, there is no way we can check for validity
+of those. The intention of this check was to prevent spams. We will remove
+this check and rely on economic incentives so that the leader can perform
+appropriate checks.
+
+The vote only execution will operate exclusively on replicated vote states
+stored outside the accounts DB, so vote only execution and full execution can
+happen asynchronously in any order. The vote authority of each vote account
+will be copied from accounts DB at the beginning of each epoch, this means
+in the future vote authority change will take two epochs instead of one
+epoch.
+
+### Replay the full block on selected forks asynchronously
+
+There is no protocol enforced order of block replay for various validator
+implemenations, new vote transactions could be sent when the vote only hash
+or replay tip hash changes.
+
+Once a validator has determined the fork it will vote on, it can prioritize
+replaying blocks on the selected fork. The replay process is the same as today,
+all transactions (vote and non-vote) will be executed to determine the final
+bankhash. The computed bankhash will be attached to vote instructions. So we
+can still detect non-determinism (same of set of instructions leading to
+different results) like today, only that maybe at a later time.
+
+To guarantee the blockchain will halt when full replay is having problems, we
+propose:
+
+1. If full replay is behind vote only replay by more than 1/2 epoch or vice
```

Comment:
re: "vice versa" can full replay run before vote only replay? I think we should ensure that full replay cannot run on a block that has not been vote only replayed

Reply:
There is no reason a full replay can't start before the bank is vote only frozen, they write into completely different set of states.

Reply:
makes sense, we should outline the interaction here:
> Once a validator has determined the fork it will vote on, it can prioritize
replaying blocks on the selected fork

maybe point out that if we haven't completed vote replay we replay all available forks equally. also outline what happens in the case that we're full replaying a fork but end up vote replaying and voting on a different fork.

Reply:
I think what you described is reasonable, but I prefer not to specify in what order the full replay should happen in the protocol. Anza and Jump may decide to use different fork weight algorithm to pick which block to send to full replay next, I think the protocol would still work. And from what I heard, playing multiple forks at the same time or playing into the distant future is also a possibility.